rpc-broker/src/msg.c

```diff
@@ -0,0 +1,245 @@
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+*/
+
+#include "rpc-broker.h"
+
+
+/*
+ * All requests are handled by this function whether they are raw requests
+ * coming from port 5555 or Websocket requests from 8080.  For every policy
+ * rule listed either in /etc/rpc-broker.policy file or the domain-specific
+ * rules listed in xenclient database are passed to `filter` to determine
+ * whether or not to be dropped or passed through.
+ */
+int broker(struct dbus_message *dmsg, int domid)
+{
+    if (!dmsg) {
+        DBUS_BROKER_WARNING("Invalid args to broker-request %s", "");
+        return 1;
+    }
+
+    if (!dbus_broker_policy) {
+        DBUS_BROKER_WARNING("No policy in place %s", "");
+        return 1;
+    }
+
+    int policy = 0;
+
+    struct etc_policy etc = dbus_broker_policy->etc;
+
+    int i, j;
+    for (i=0; i < etc.count; i++)
+        policy = filter(&(etc.rules[i]), dmsg, domid);
+
+    int domains = dbus_broker_policy->domain_number;
+
+    for (i=0; i < domains; i++) {
+        if (dbus_broker_policy->domains[i].domid == domid) {
+
+            struct domain_policy domain = dbus_broker_policy->domains[i];
+
+            for (j=0; j < domain.count; j++)
+                policy = filter(&(domain.rules[j]), dmsg, domid);
+
+            break;
+        }
+    }
+
+    if (verbose_logging) {
+        char req_msg[1024];
+
+        snprintf(req_msg, 1023, "Dom: %d [Dest: %s Path: %s Iface: %s Meth: %s]",
+                          domid, dmsg->destination, dmsg->path,
+                                 dmsg->interface, dmsg->member);
+
+        if (policy == 0)
+            DBUS_BROKER_WARNING("%s <%s>", req_msg, "Dropped request");
+        else
+            DBUS_BROKER_EVENT("%s <%s>", req_msg, "Passed request");
+    }
+
+    return policy;
+}
+
+/*
+ * This is an opaque function to pass raw dbus-bytes back in forth between
+ * sender and receiver.  Its opaque in the sense that the function is unaware
+ * who the client is and who the server is.  There are just sender and receiver
+ * syscalls being made over port 5555.
+ */
+int exchange(int rsock, int ssock, int domid,
+             ssize_t (*rcv)(int, void *, size_t, int),
+             ssize_t (*snd)(int, const void *, size_t, int))
+{
+    int total = 0;
+    char buf[DBUS_MSG_LEN] = { 0 };
+
+    while ( 1 ) {
+
+        int rbytes = rcv(rsock, buf, DBUS_MSG_LEN, 0);
+
+        if (rbytes <= 0)
+            break;
+
+        if (rbytes > DBUS_COMM_MIN) {
+
+            struct dbus_message dmsg;
+            int len = dbus_message_demarshal_bytes_needed(buf, rbytes);
+
+            if (len == rbytes) {
+
+                if (convert_raw_dbus(&dmsg, buf, len) < 1)
+                    return -1;
+
+                //if (broker(&dmsg, domid) < 1)
+                //    return -1;
+            }
+
+            if (!dom0) {
```

Comment:
Is this debug logic? If so, probably be good to wrap with a "#define DEBUG" so that it can be compiled out in production.